### PR TITLE
test: Add Integration Testing

### PR DESCRIPTION
**Pull request by suneettipirneni**
**Please describe the changes this PR makes and why it should be merged:**

Doing integration tests on a discord library is not straightforward. There was a couple way to approach the situation:

- Stub out parts of the library that rely on the API
- Mock the API itself

While the first choice is faster in terms of performance it's also quite problematic. Firstly, stubbing out parts of the library that hit the API would be a lot of upkeep in addition to writing the tests themselves. Secondly, stubbing out the code can give a lot of false positives when testing, since you're not testing against the true source code anymore.

Because of those reasons I decided to mock the API + Gateway instead. I created [a package](https://github.com/suneettipirneni/discord-mock-server) that mocks both the gateway and API.

Using the mock server allows us to not have to modify any of the discord.js source code while also allowing us to manually trigger gateway events, and stub out API calls.

Currently I implemented tests for action handlers that handle messages and interactions. I was able to achieve full code coverage on both gateway handlers using this strategy. In the future I plan to add more tests for gateway handlers.

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating

### CODE REVIEW

File: packages/discord.js/.vscode/settings.json
Comment (iCrawl) [inline packages/discord.js/.vscode/settings.json:3]: No.

**Review by kyranet — CHANGES_REQUESTED** [review]
One thing for now, the PR is large and will take a while to get reviewed.

**Comment by kyranet** [comment]
Another thing, I believe this is not exactly what we have in mind, we shouldn't need to mock Discord in order to make sure the library works, instead, we should only do the following things:

- `@discordjs/rest`'s tests.
- `@discordjs/ws`'s code + tests.

Once we have both REST and WS tested using their domain-specific tools, testing discord.js itself is just a matter of testing two things:

- **Unit testing**: see that all structures build just fine and that getters return the right values.
- **Integration testing**: see that the REST methods call the right methods with the right data. This can be done easily with HTTP interceptors, something we'll know how to do once we have REST's tests working with undici.

As of right now, none of the previous tasks are done, so I believe implementing integration tests at this moment will only require us to rewrite a lot of code in the future, specially since we will probably want to run unit and integration tests separately, since the latter are far more expensive. There's just too much to figure out right now.

I'll leave this to the other maintainers to decide, I'm just voicing the concerns I have based on previous experience in similar projects.

cc: @discordjs/the-big-4

File: packages/discord.js/test/MockClient.js
Comment (vladfrangu) [reply packages/discord.js/test/MockClient.js:21]: you mean once?

**Pull request closed without merging**
